slides/advanced-02-racing.qmd

```diff
@@ -0,0 +1,349 @@
+1. Eliminate poor candidates
+1. Repeat with next resample. 
+
+This can result in fitting a small number of models. 
+
+
+## Discarding Candidates
+
+How to we eliminate tuning parameter combinations? 
+
+There are a few methods to do so. We'll use one based on analysis of variance (ANOVA). 
+
+_However_... there is typically a large effect due to resamples in the data. 
+
+## Resampling Results (Non-Racing)
+
+:::: {.columns}
+
+::: {.column width="50%"}
+
+
+```{r}
+#| label: race-data-comp
+#| echo: false
+#| out-width: 60%
+#| fig-align: center
+#| fig-width: 5
+#| fig-height: 5
+
+# Simulate some resamples for candidate models
+num_resamples <- 10
+race_example <- crossing(model = paste("candidate", 1:2), resample = 1:num_resamples)
+candidates <- tibble(model = paste("candidate", 1:2), mean = c(4, 2))
+set.seed(937)
+resamples <- tibble(resample = 1:num_resamples, effect = rnorm(num_resamples, sd = 2))
+race_example <-
+  race_example %>%
+  full_join(candidates, by = "model") %>%
+  full_join(resamples, by = "resample") %>%
+  mutate(
+    error = mean + effect + rnorm(n(), sd = 1),
+    resample = format(resample)
+  ) %>%
+  select(-mean, -effect)
+
+
+race_table <- 
+  race_example %>%
+  mutate(model = gsub(" ", "_", model)) %>%
+  pivot_wider(id_cols = c(resample), names_from = model, values_from = error)
+
+race_cor <- cor(race_table[, -1], method = "spearman")[1,2]
+
+get_ci <- function(ind) {
+  dat <- race_table[1:ind,]
+  t_test <- t.test(dat$candidate_1, dat$candidate_2, paired = TRUE)
+  t_test <- tidy(t_test, conf)
+  t_test %>% 
+    select(difference = estimate, lower = conf.low, upper = conf.high) %>% 
+    mutate(`number of resamples` = ind)
+}
+
+race_ci <- map_dfr(2:10, get_ci) 
+
+```
+
+Here are some realistic (but simulated) examples of two candidate models. 
+
+An error estimate is measured for each of 10 resamples. 
+
+ - The lines connect resamples. 
+
+There is usually a significant resample-to-resample effect (rank corr: `r round(race_cor, 2)`). 
+
+:::
+
+::: {.column width="50%"}
+
+
+```{r}
+#| label: race-data
+#| echo: false
+#| out-width: 100%
+#| fig-align: center
+#| fig-width: 5
+#| fig-height: 5
+
+
+race_example %>%
+  ggplot(aes(x = model, y = error, group = resample, col = factor(resample))) +
+  geom_point(show.legend = FALSE) +
+  geom_line(alpha = 1 / 2, show.legend = FALSE) +
+  labs(x = NULL, y = "model error")
+```
+
+:::
+
+::::
+
+
+## Are Candidates Different?
+
+One way to evaluate these models is to do a paired t-test
+ 
+ - or a t-test on their differences matched by resamples
+
+With $n = 10$ resamples, the confidence interval is (`r signif(race_ci$lower[9], 2)`, `r signif(race_ci$upper[9], 2)`), indicating that candidate number 2 has smaller error. 
+
+What if we were to compare each model candidate to the current best at each resample? 
+
+One shows superiority when `r which.max(race_ci$lower > 0)` resamples have been evaluated.
+
+
+## Evaluating Differences in Candidates
```

Comment:
I liked how the plots in your conf talk from 2021 showed the process per iteration/resample, i.e. that "evaluating the differences between candidates" and "candidates are being discarded" were being shown together.

For anyone not Max: from about the 5:30 mark of https://posit.co/resources/videos/whats-new-in-tidymodels/

Reply:
I'll add an issue to make this same animation for the posit conf version. (not sure that I have enough time currently)